src/coredump/coredump.c

```diff
@@ -1328,6 +1329,169 @@ static int gather_pid_metadata_from_procfs(struct iovec_wrapper *iovw, Context *
+                .uid = context->uid,
+                .gid = context->gid,
+        };
+        int r;
+
+        r = get_namespace_leader(context->pid, NAMESPACE_PID, &pid);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get container leader PID: %m");
+
+        if (socketpair(AF_UNIX, SOCK_DGRAM, 0, pair) < 0)
+                return -errno;
+
+        /* We need to translate the PID, UID, and GID of the crashing process
+         * to the container's namespaces. Do this by sending an SCM_CREDENTIALS
+         * message on a socket pair, and read the result when we join the
+         * container. The kernel will perform the translation for us. */
+        r = send_ucred_and_one_fd(pair[0], &ucred, STDIN_FILENO);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to send metadata to container: %m");
+
+        r = namespace_open(pid, &pidnsfd, &mntnsfd, &netnsfd, &usernsfd, &rootfd);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to join namespaces of PID " PID_FMT ": %m", pid);
+
+        r = namespace_fork("(sd-coredumpns)", "(sd-coredump)", NULL, 0, FORK_RESET_SIGNALS|FORK_DEATHSIG,
+                           pidnsfd, mntnsfd, netnsfd, usernsfd, rootfd, &child);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to fork into namespaces of PID " PID_FMT ": %m", pid);
+        if (r == 0) {
+                _cleanup_close_ int input_fd = -EBADF;
+                char pidstr[DECIMAL_STR_MAX(pid_t)];
+                char uidstr[DECIMAL_STR_MAX(uid_t)];
+                char gidstr[DECIMAL_STR_MAX(gid_t)];
+
+                pair[0] = safe_close(pair[0]);
+
+                r = receive_ucred_and_one_fd(pair[1], &ucred, &input_fd);
+                if (r < 0)
+                        _exit(EXIT_FAILURE);
+
+                xsprintf(pidstr, PID_FMT, ucred.pid);
+                xsprintf(uidstr, UID_FMT, ucred.uid);
+                xsprintf(gidstr, GID_FMT, ucred.gid);
+
+                r = move_fd(input_fd, STDIN_FILENO, false);
+                if (r < 0)
+                        _exit(EXIT_FAILURE);
+
+                const char *cmd[] = {
+                        "/usr/lib/systemd/systemd-coredump",
+                        pidstr,
+                        uidstr,
+                        gidstr,
+                        context->meta[META_ARGV_SIGNAL],
+                        context->meta[META_ARGV_TIMESTAMP],
+                        context->meta[META_ARGV_RLIMIT],
+                        context->meta[META_ARGV_HOSTNAME],
+                        NULL,
+                };
+
+                execv(cmd[0], (char * const *) cmd);
+                _exit(EXIT_FAILURE);
+        }
+
+        pair[1] = safe_close(pair[1]);
+
+        r = wait_for_terminate_and_check("(sd-coredumpns)", child, 0);
+        if (r < 0) {
+                return log_debug_errno(r, "Failed to wait for child process to terminate: %m");
+        }
```

Comment:
Pretty sure you can replace this with `FORK_WAIT` which does the same internally in `namespace_fork()`